### PR TITLE
opt: improve error message for failing to force a partial index

### PR DESCRIPTION
**Pull request by mgartner**
This commit expands the error message when a user tries to force a
partial index that cannot satisfy the query. This error is emitted when
the query filter does not imply the partial index predicate.

Release note: None

### CODE REVIEW

File: pkg/sql/opt/exec/execbuilder/relational.go [hunk 1]
```diff
@@ -457,6 +457,10 @@ func (b *Builder) scanParams(
 		var err error
 		if idx.IsInverted() {
 			err = fmt.Errorf("index \"%s\" is inverted and cannot be used for this query", idx.Name())
+		} else if _, isPartial := idx.Predicate(); isPartial {
+			err = fmt.Errorf(
+				"index \"%s\" is a partial index and cannot be used for this query because the query filter does not imply the predicate",
```
Comment (mgartner) [inline pkg/sql/opt/exec/execbuilder/relational.go:462]: Too wordy?

**Review by RaduBerinde — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51581#-:-MCddDK0E8KYotRSIEHb:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @mgartner)

___
*[pkg/sql/opt/exec/execbuilder/relational.go, line 462 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51581#-MCYajNsnykjERtH16Oc-r1-462:-MCdc9HZD1IM_cP2yFdm:b-wjj6lw) ([raw file](https://github.com/cockroachdb/cockroach/blob/704619482e41926b49e4044822f606b5447141bf/pkg/sql/opt/exec/execbuilder/relational.go#L462)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Too wordy?
</blockquote></details>

Users might not know what "the predicate" refers to. Even "index predicate" might not be enough.
Maybe keep it more vague and say "is a partial index that is unsuitable for executing this query" (or maybe "was found unsuitable")

**Review by rytaft — APPROVED** [review]
:lgtm: 

> Reviewed 2 of 2 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51581#-:-MCgbIR45pYBA3SfqdVA:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @mgartner)

___
*[pkg/sql/opt/exec/execbuilder/relational.go, line 462 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51581#-MCYajNsnykjERtH16Oc-r1-462:-MCgajx-7ZG0AZWSf-GF:b-8nqd25) ([raw file](https://github.com/cockroachdb/cockroach/blob/704619482e41926b49e4044822f606b5447141bf/pkg/sql/opt/exec/execbuilder/relational.go#L462)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Users might not know what "the predicate" refers to. Even "index predicate" might not be enough.
Maybe keep it more vague and say "is a partial index that is unsuitable for executing this query" (or maybe "was found unsuitable")
</blockquote></details>

Slightly less vague alternative: "is a partial index that does not contain all the rows needed to execute this query". Feel free to ignore if you're happy with it as-is.

**Review by mgartner — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51581)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @RaduBerinde and @rytaft)

___
*[pkg/sql/opt/exec/execbuilder/relational.go, line 462 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51581#-MCYajNsnykjERtH16Oc-r1-462:-MCh72mf325s6RLM4BK_:blridlz) ([raw file](https://github.com/cockroachdb/cockroach/blob/704619482e41926b49e4044822f606b5447141bf/pkg/sql/opt/exec/execbuilder/relational.go#L462)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Slightly less vague alternative: "is a partial index that does not contain all the rows needed to execute this query". Feel free to ignore if you're happy with it as-is.
</blockquote></details>

There's more to it than containment of the columns. The partial index predicate has to be implied by the filter. It's complicated enough that any single-sentence explanation of this is bound to be vague. My first version tried to be more specific but @raduberinde is right that it was confusing what a "predicate" refers to.

But maybe something like "index abc is a partial index and cannot be used because its WHERE clause is not implied by the query's WHERE clause".

**Comment by rytaft** [comment]
___
*[pkg/sql/opt/exec/execbuilder/relational.go, line 462 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51581#-MCYajNsnykjERtH16Oc-r1-462:-MChIP0D6BFljIWRDa3k:bxkuksp) ([raw file](https://github.com/cockroachdb/cockroach/blob/704619482e41926b49e4044822f606b5447141bf/pkg/sql/opt/exec/execbuilder/relational.go#L462)):*
> There's more to it than containment of the columns

I suggested containment of rows, not columns

But like I said, feel free to ignore

**Review by mgartner — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51581)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @rytaft)

___
*[pkg/sql/opt/exec/execbuilder/relational.go, line 462 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51581#-MCYajNsnykjERtH16Oc-r1-462:-MChR82T9DXLPnAG5qbT:b-xnvbp) ([raw file](https://github.com/cockroachdb/cockroach/blob/704619482e41926b49e4044822f606b5447141bf/pkg/sql/opt/exec/execbuilder/relational.go#L462)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

> There's more to it than containment of the columns

I suggested containment of rows, not columns

But like I said, feel free to ignore
</blockquote></details>

Doh. Read your message too fast. I love it. Done!

**Review by rytaft — APPROVED** [review]
> Reviewed 2 of 2 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51581)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale)

___
*[pkg/sql/opt/exec/execbuilder/relational.go, line 462 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51581#-MCYajNsnykjERtH16Oc-r1-462:-MCi8rQ239aWDYsZwNPP:b4c9lyr) ([raw file](https://github.com/cockroachdb/cockroach/blob/704619482e41926b49e4044822f606b5447141bf/pkg/sql/opt/exec/execbuilder/relational.go#L462)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Doh. Read your message too fast. I love it. Done!
</blockquote></details>

:thumbsup:

**Comment by RaduBerinde** [comment]
___
*[pkg/sql/opt/exec/execbuilder/relational.go, line 462 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51581#-MCYajNsnykjERtH16Oc-r1-462:-MCiI3xx58yGlOdm05gR:bafstpj) ([raw file](https://github.com/cockroachdb/cockroach/blob/704619482e41926b49e4044822f606b5447141bf/pkg/sql/opt/exec/execbuilder/relational.go#L462)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

:thumbsup: 
</blockquote></details>

Very nice wording!
